src/tools/rados/rados.cc

```diff
@@ -1142,6 +1142,232 @@ static int do_cache_flush_evict_all(IoCtx& io_ctx, bool blocking)
+    for (buf = obj_list_buf, next_obj_cursor = NULL; ; buf = NULL) {
+      char *obj_name, delim = '\n';
+      if ((obj_name = strtok_r(buf, &delim, &next_obj_cursor)) == NULL) {
+	break;
+      }
+      string obj_name_str(obj_name);
+      if (fast_flush_one_obj(*(ffd->io_ctx), obj_name_str) < 0) {
+	if ((++failed) % 1000 == 0) {
+	  cerr << "thread: " << pthread_self()
+	       << ", evict fail count: " << failed << std::endl;
+	}
+      } else {
+	if ((++passed) % 1000 == 0) {
+	  cerr << "thread: " << pthread_self()
+	       << ", evict pass count: " << passed << std::endl;
+	}
+      }
+      if ((++processed) % 1000 == 0) {
+	cerr << "thread: " << pthread_self()
+	     << ", processed count: " << processed << std::endl;
+      }
+    }
+  }
+
+  delete obj_list_buf;
+  cout << "thread: " << pthread_self()
+       << ", processed: " << processed
+       << ", passed: " << passed
+       << ", failed: " << failed
+       << std::endl;
+
+  fast_flush_thread_cnt.inc();
+  pthread_mutex_lock(&fast_flush_thread_mutex);
+  total_processed += processed;
+  total_passed += passed;
+  total_failed += failed;
+  pthread_cond_signal(&fast_flush_thread_cond);
+  pthread_mutex_unlock(&fast_flush_thread_mutex);
+  return NULL;
+}
+
+IoCtx *fast_flush_io_ctx(const char *pool_name)
+{
+  int ret;
+  Rados *rados = new (std::nothrow) Rados;
+  if (ret = rados->init_with_context(g_ceph_context)) {
+     cerr << "RADOS initialize error: " << ret << std::endl;
+     return NULL;
+  }
+  if (ret = rados->connect()) {
+     cerr << "RADOS connect error: " << ret << std::endl;
+     return NULL;
+  }
+  IoCtx *io_ctx = new (std::nothrow) IoCtx;
+  if ((ret = rados->ioctx_create(pool_name, *io_ctx)) < 0) {
+    cerr << "Pool open error: " << pool_name << ": "
+	 << cpp_strerror(ret) << std::endl;
+    return NULL;
+  }
+  return io_ctx;
+}
+
+static int do_cache_fast_flush(const char *obj_list_fn,
+			       int num_threads, const char *pool_name)
+{
+  int fd;
+  if ((fd = ::open(obj_list_fn, O_RDONLY)) == -1) {
+    cerr << "Error opening file: " << obj_list_fn << std::endl;
+    return -errno;
+  }
+
+  struct stat st_buf;
+  if (::fstat(fd, &st_buf) == -1) {
+    return -errno;
+  }
+  if (st_buf.st_size > OBJ_LIST_SIZE_MAX) {
+    cerr << "Input file too big, " << st_buf.st_size << std::endl;
+    return -EFBIG;
+  }
+
+  vector<off_t> *offsets = split_obj_list(st_buf.st_size, num_threads);
+  if (offsets->empty()) {
+    cerr << "Error splitting file: " << obj_list_fn << std::endl;
+    return -EINVAL;
+  }
+
+  if (num_threads > (offsets->size() - 1)) {
+    num_threads = offsets->size() - 1;
+  }
+  cout << "Will spawn " << num_threads
+       << " threads to flush from " << obj_list_fn
+       << std::endl;
+
+  total_processed = total_passed = total_failed = 0;
+
+  pthread_t *thread_ids = new (std::nothrow) pthread_t[num_threads];
+  pthread_attr_t attr;
+  pthread_attr_init(&attr);
+
+  pthread_mutex_init(&fast_flush_thread_mutex, NULL);
```

Comment:
we can use Mutex class, there is no functionality difference, will get some additional debugging help if needed.